compiler/backend/cgir2.nim

```diff
@@ -0,0 +1,349 @@
+    cnkFToI, cnkFToU, cnkUToF, cnkIToF
+    cnkSizeof, cnkAlignof, cnkOffsetof
+    cnkUnlikely
+    cnkEmit, cnkAsm
+    cnkStmtList
+    cnkBlock
+    cnkScope
+    cnkTry
+    cnkIf
+    cnkDispatch
+    cnkTarget
+    cnkWhile
+    cnkBreak, cnkRaise, cnkReturn, cnkUnreachable
+    cnkDef
+    cnkAsgn, cnkStore
+    cnkDrop
+    cnkCheckedCall
+    cnkCheckedCallAsgn
+    cnkTailCall
+    cnkCall
+    cnkConstr, cnkRecConstr, cnkFieldInit
+    cnkParams, cnkParam
+    cnkProcImp, cnkProcDef, cnkProcExp
+    cnkGlobalImp, cnkGlobalDef, cnkGlobalExp
+    cnkVoidTy, cnkBoolTy, cnkCharTy, cnkUIntTy, cnkIntTy, cnkFloatTy
+    cnkOpaqueTy, cnkStructTy, cnkUnionTy, cnkArrayTy
+    cnkPtrTy, cnkProcTy
+    cnkField
+    cnkVarargs
+
+  CgTypeKind* = range[cnkVoidTy..cnkProcTy]
+
+  CgNode* = object
+    ## A single packed node in an abstract syntax tree.
+    kindInfo*: uint32
+      ## 8-bit kind; 24-bit info index (0 meaning 'none')
+    val*: uint32
+      ## for terminals, meaning depends on the kind
+      ## for non-terminals, the number of child nodes
+
+  Ast* = seq[CgNode]
+    ## Packed storage for AST.
+  NodeIndex* = distinct uint32
+    ## Index of a node in an `Ast`.
+  Datum* = distinct uint32
+    ## ID of an anonymous constant.
+
+  SourceLoc* = object
+    ## Source location information.
+    line*: uint16   ## line number; '0' means invalid
+    column*: uint16 ## column number. '0' means "full line"
+    file*: StringId ## full path of the source file
+
+  AsmMode* = enum
+    ## What an inline asm statement represents.
+    asmMsvc ## MSVC inline assembly syntax
+    asmGnu  ## GNU inline assembly syntax
+    asmJs   ## inline JavaScript
+
+  CgStorage* {.pure.} = enum
+    Normal ## mutable global storage
+    Thread ## thread-local storage
+    Const  ## constant global storage
+
+  CgProcAttrib* {.pure.} = enum
+    Inline   ## try to inline the body where possible
+    NoInline ## never inline the body at a callsite
+
+  CgParamAttrib* {.pure.} = enum
+    ## Parameter attributes.
+    NoAlias ## C-specific attribute; may be ignored
+
+  CgLocAttrib* {.pure.} = enum
+    ## Attributes for locals, globals, and fields.
+    NoAlias  = 0 ## C-specific attribute; may be ignored
+    Register = 1 ## C-specific attribute; may be ignored
+    Volatile = 2 ## C-specific attribute; may be ignored
+
+  CgCallConv* {.pure.} = enum
+    ## Calling convention.
+    Default   ## decided by the implementation
+    Nimcall   ## calling convention for NimSkull procedures
+    Stdcall,  ## C stdcall
+    Cdecl,    ## C cdecl
+    Safecall, ## C safecall
+    Syscall,  ## C syscall
+    Fastcall  ## C fastcall
+
+  CgModule* = object
+    ## A CGIR module using an in-memory representation meant for
+    ## easy processing.
+    tast*: Ast
+      ## all AST for type bodies
+    ast*: Ast
+      ## all AST for everything that's not types
+    data*: Store[Datum, NodeIndex]
+      ## anonymous constants represented as construction expressions
```

Comment:
```suggestion
      ## anonymous constants represented as construction expressions
      ## constant datum (id) -> index of corresponding declaration in `ast`
```

I haven't read that far yet, but this is my guess of where it's looking things up.

Reply:
I've only improved the comment for now.

Just adding a dedicated constant declaration syntax wouldn't make the AST self-contained at this point, as there's still the "loose" storage types of, and I'd rather combine these changes with adding a (somewhat) proper transportation format (plus the associated reader/writer and integration into the compiler thereof).

Also, to maybe clarify, with self-contained I mean that the AST contains all information necessary to process it / work with it (without having to store any of the lookup tables - except the out-of-band values - currently part of `CgModule` alongside it).